mining/src/manager.rs

```diff
@@ -0,0 +1,178 @@
+    }
+
+    pub fn get_block_template(&self, miner_data: &MinerData) -> MiningManagerResult<BlockTemplate> {
+        let cache_read = self.block_template_cache.upgradable_read();
+        let immutable_template = cache_read.get_immutable_cached_template();
+
+        // We first try and use a cached template if not expired
+        if let Some(immutable_template) = immutable_template {
+            drop(cache_read);
+            if immutable_template.miner_data == *miner_data {
+                return Ok(immutable_template.as_ref().clone());
+            }
+            // Miner data is new -- make the minimum changes required
+            // Note the call returns a modified clone of the cached block template
+            let block_template = self.block_template_builder.modify_block_template(miner_data, immutable_template)?;
+
+            // No point in updating cache since we have no reason to believe this coinbase will be used more
+            // than the previous one, and we want to maintain the original template caching time
+            return Ok(block_template);
+        }
+
+        // Rust rewrite:
+        // We avoid passing a mempool ref to blockTemplateBuilder by calling
+        // mempool.BlockCandidateTransactions and mempool.RemoveTransactions here.
+        // We remove recursion seen in blockTemplateBuilder.BuildBlockTemplate here.
+        let mut cache_write = RwLockUpgradableReadGuard::upgrade(cache_read);
+        loop {
+            let transactions = self.block_candidate_transactions();
+            match self.block_template_builder.build_block_template(miner_data, transactions) {
+                Ok(block_template) => {
+                    let block_template = cache_write.set_immutable_cached_template(block_template);
+                    return Ok(block_template.as_ref().clone());
+                }
+                Err(BuilderError::ConsensusError(RuleError::InvalidTransactionsInNewBlock(invalid_transactions))) => {
+                    let mut mempool_write = self.mempool.write();
+                    let removal_result = invalid_transactions.iter().try_for_each(|(x, _)| mempool_write.remove_transaction(x, true));
+                    drop(mempool_write);
+                    if let Err(err) = removal_result {
+                        // Original golang comment:
+                        // mempool.remove_transactions might return errors in situations that are perfectly fine in this context.
+                        // TODO: Once the mempool invariants are clear, this might return an error:
+                        // https://github.com/kaspanet/kaspad/issues/1553
+                        error!("Error from mempool.remove_transactions: {:?}", err);
+                    }
+                }
+                Err(err) => {
+                    return Err(err)?;
+                }
+            }
+        }
+    }
+
+    pub(crate) fn block_candidate_transactions(&self) -> Vec<MutableTransaction> {
```

Comment:
Add a TODO suggesting to check the cost of cloning all mempool transactions (if #mempool txs >> #txs which fit in a block, then most of these clones will not be used).